### PR TITLE
add model_diff() function

### LINKED ISSUE #324: model_diff function
# Summary

As a user, I would like to be able to see a rendered diff between two model files, either printed to the console or rendered in Rmd files. This should default to diff'ing the control streams for NONMEM and optionally diff other model files for Stan models.

# Tests
- tests/testthat/test-model-diff.R
  - model_diff.bbi_nonmem_model happy path based_on
  - model_diff.bbi_nonmem_model happy path .mod2 arg
  - model_diff.bbi_nonmem_model errors with no based_on
  - model_diff.bbi_nonmem_model errors with multiple based_on

### PR DESCRIPTION
**Pull request by seth127**
closes #324

### CODE REVIEW

**Comment by seth127** [comment]
# For Review
This is largely an interactive function, so unit tests only go so far in terms of testing it. Below are some screen shots and some attached files to can be used for manual testing.

## Rmd rendering

To test knitting an `.Rmd` file, pull this branch and open `tests/testthat/diff-test.Rmd` and knit it. You can also just download `tests/testthat/diff-test.html` to see the rendered version. **NOTE: these files need to be deleted before this branch merges.**

Here are screenshots of the rendered `.Rmd`
Screen Shot 2021-03-02 at 2 12 22 PM
Screen Shot 2021-03-02 at 2 12 29 PM

## Calling from the console

With no args passed, prints to the console:
Screen Shot 2021-03-02 at 2 04 45 PM

With `.viewer = TRUE` renders in viewer window and holds the console:
Screen Shot 2021-03-02 at 2 04 14 PM

## Running `.Rmd` blocks interactively

With no args it prints to the "console" in the `.Rmd` (with or without `results = "asis"`)
Screen Shot 2021-03-02 at 2 00 46 PM
Screen Shot 2021-03-02 at 2 01 00 PM

With `.viewer = TRUE` it renders in the viewer and holds the "console", i.e. holds the blocks execution until you hit enter. (with or without `results = "asis"`)
Screen Shot 2021-03-02 at 2 01 29 PM
Screen Shot 2021-03-02 at 2 01 58 PM

**Comment by timwaterhouse** [comment]
Looks good, thanks Seth. A couple of small requests:
* Is there any way to change the headers ("model_A" and "model_B") to the model names?
* Prompting to press enter is a minor annoyance. Can we bypass that?

**Comment by seth127** [comment]
> Is there any way to change the headers ("model_A" and "model_B") to the model names?

I'll look into this. That would be nice, but I'm not actually sure where `diffobj` is pulling that from...

> Prompting to press enter is a minor annoyance. Can we bypass that?

I messed with this for awhile and the current situation is that it only happens when you pass `.viewer = TRUE` (it defaults to `FALSE`). Otherwise, it just prints to the console. Unfortunately, I can't figure out any way to get it to show in the viewer window but _not_ give you this annoying "Press ENTER" prompt. I'm sure it's possible, but I spent more time than I'd like to admit on it and couldn't figure it out.

File: R/tags-diff.R
Comment (seth127) [inline R/tags-diff.R:53]: this is unrelated to this PR, but I realized when writing `model_diff()` that I should check this here too.

File: R/model-diff.R
Comment (callistosp) [inline R/model-diff.R:134]: do we really want it to silently fail if the user does `.viewer=T` but doesn't include `results='asis'`? Should we print a warning to the output like "to include output in knitted file, set chunk option to results=asis"?

**Review by callistosp — CHANGES_REQUESTED** [review]
looks good @seth127, I have a couple change requests and questions in here. I can approve once these are addressed

**Comment by seth127** [comment]
@callistosp thanks for the feedback, I'll look at this this afternoon. 

@timwaterhouse I figured out a solution to your question, but it's a bit hacky. I posted a question on the package repo to see if the maintainer has any advice: https://github.com/brodieG/diffobj/issues/154

File: R/aaa.R
Comment (seth127) [reply R/aaa.R:233]: no, but diffobj will print a "diff" that at the top says "no differences" or something like that. I don't really think we need to explicitly catch that and raise an error, do you?

File: R/model-diff.R
Comment (seth127) [reply R/model-diff.R:24]: bingo https://github.com/metrumresearchgroup/bbr/issues/341#issuecomment-789067691

File: R/aaa.R
Comment (callistosp) [reply R/aaa.R:233]: agreed

File: R/model-diff.R
Comment (seth127) [reply R/model-diff.R:134]: good call. I'll add a warning if someone is knitting and has `.viewer=T` in there

Comment (seth127) [reply R/model-diff.R:134]: Check it:
Screen Shot 2021-03-03 at 3 28 57 PM

File: R/aaa.R
Comment (seth127) [reply R/aaa.R:233]: Screen Shot 2021-03-03 at 3 26 56 PM

File: R/model-diff.R
Comment (seth127) [reply R/model-diff.R:134]: I also added a warning for trying to use this in non-HTML knitting. See the bottom of this comment https://github.com/metrumresearchgroup/bbr/pull/342#issuecomment-789150414

File: tests/testthat/test-model-diff.R
Comment (seth127) [reply tests/testthat/test-model-diff.R:52]: yea we don't need to test that, but just so you know, the call with go:
```r
> model_diff()
>> model_diff_get_comp()
>>> check_yaml_in_sync()
>>>> check_model_object()
stop(glue("Must pass a model object, but got object of class: `{paste(class(.mod), collapse = ', ')}`"))
```

**Comment by seth127** [comment]
@timwaterhouse @callistosp I have some misgivings about [this renaming thing](https://github.com/metrumresearchgroup/bbr/blob/feat/model_diff/R/model-diff.R#L162) (described in that issue I linked in the previous comment) but ultimately I think it's better and probably totally safe. My vote is to go ahead and merge it and we can make a patch to change it if the package developer comes back with some very different (or adds that feature).

**Comment by timwaterhouse** [comment]
> @timwaterhouse @callistosp I have some misgivings about [this renaming thing](https://github.com/metrumresearchgroup/bbr/blob/feat/model_diff/R/model-diff.R#L162) (described in that issue I linked in the previous comment) but ultimately I think it's better and probably totally safe. My vote is to go ahead and merge it and we can make a patch to change it if the package developer comes back with some very different (or adds that feature).

Sounds good to me, thanks.

File: R/model-diff.R
Comment (callistosp) [inline R/model-diff.R:88]: perfect

**Review by callistosp — APPROVED** [review]
LGTM

**Review by callistosp — DISMISSED** [review]

**Comment by seth127** [comment]
Merging this with my privileges because @callistosp approved it yesterday and all that's changed is that we merged its base branch into develop and I made the tiny refactor recommended by the `diffobj` developer in [this comment](https://github.com/brodieG/diffobj/issues/154#issuecomment-790625545).

Screen Shot 2021-03-04 at 10 09 52 AM

**Comment by callistosp** [comment]
> Merging this with my privileges because @callistosp approved it yesterday and all that's changed is that we merged its base branch into develop and I made the tiny refactor recommended by the `diffobj` developer in [this comment](https://github.com/brodieG/diffobj/issues/154#issuecomment-790625545).

a gross misuse of power

**Pull request merged**
